improve identification of imaginary roots for CRootOf instances

smichr opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests .-->

#14293 #14291 

#### Brief description of what is fixed or changed

fixes #14615

Fixes #14291. Because not all imaginary intervals were being identified, some imaginary intervals found their way into code that attempts to refine the interval until both bottom left and upper right were modified. An imaginary interval containing an x-value of 0 at one of the corners will never refine that corner and an infinite loop ensues.

The imaginary identification has now been made robust allowing intervals to be clearly identified as having a real part or being purely imaginary. This allows an `_eval_is_imaginary` routine to be added to CRootOf: so where `CRootOf(x**5+x-1,3).is_imaginary` formerly gave None, it now gives False since it is not real and not imaginary. The cost of doing this is extra refinement at instantiation.

#### Other comments

The sorting of complex roots has been abandoned until a method of determining when two roots share a common real value. (See discussion at #14293.) Instead, we settle for the invariant that the real roots come first followed by complex pairs -- in no guaranteed order -- but with the one having a negative real part coming before the positive real part. So the conjugate of a non-real root will always be an index of +/-1 from a root's index:
```
>>> CRootOf(x**4 + x + 1, 1).conjugate()
CRootOf(x**4 + x + 1, 0)
>>> _.conjugate()
CRootOf(x**4 + x + 1, 1)
```

The docstring of ComplexInterval has been significantly expanded.

An attempt to keep cache entries all PurePoly instances has been made and assertions inserted wherever the cache is accessed.

`_eval_Eq` has been simplified to reflect the fact that refinement has already taken place at instantiation. Also the bounds are included in the check for containment, e.g. `a <= other and other <= b` rather than `a < other and other < b`.

- [x] check that root cache is used when retrieving roots.  The only place that the root cache is being used in rootoftools is when the interval is being retreived.

smichr committed:
better handling of imaginary roots

smichr committed:
alternate evalf for CRootOf

smichr committed:
changes for new RootOf behavior

smichr committed:
expand docstring for ComplexInterval

smichr committed:
keep factors as PurePoly

smichr commented:
@jksuom , do you have a sense of whether the 'alternate evalf for CRootOf' is preferred over the method of the previous commit? If any of this is still in your memory from when we worked on this before, maybe you would have a sense of whether the condition of a root being on the corner of the bounding rectangle is something that can happen for roots that are neither real nor imaginary.

Note: the code now attempts to identify unambiguously all imaginary roots by refining until non-real and non-imaginary roots have a consistent sign for the x-boundaries. The product of x-boundaries of imaginary roots are then non-positive.

@asmeurer or @skirpichev, too, does it appear that the logic for identifying the number of imaginary roots is complete or am I missing something (still)?

skirpichev commented:
@smichr , could you describe your algorithm?

@jksuom, I would like to see your thoughts on #14293.  Is there any algorithmic way to break ties while sorting, when real parts are close (or equal)?

> the condition of a root being on the corner of the bounding rectangle is something that can happen for roots

Sure, roots could be on boundaries of the isolating rectangle.  But that's a special case, you shouldn't think that this will be eventually always after some number of refinement steps.  (But you always can check if it's the case from matadata, attached to ComplexInterval)

> Note: the code now attempts to identify unambiguously all imaginary roots by refining until non-real and non-imaginary roots have a consistent sign for the x-boundaries.

I.e. until either ax > 0 or bx < 0?  Why do you believe that this process eventually ends?  Could you prove it?

> that the logic for identifying the number of imaginary roots is complete or am I missing something (still)?

I'm not sure why you so attached to this special case and how it helps solving #14293.  Anyway, it seems for me, that you (still) assume that pure imaginary roots are present only in special factors with two terms, like ``x**2  + real```.  Which is not the case, as it was explained in #14293.  Does your ``_imag_count()`` work on ``x**4 + 10*x**2 + 1``?

jksuom commented:
> a root being on the corner of the bounding rectangle is something that can happen for roots that are neither real nor imaginary.

Yes, and I think that did occur in one of the examples.

jksuom commented:
> Is there any algorithmic way to break ties while sorting, when real parts are close (or equal)?

I would try to apply the method used in `log_to_real`. Write the root as `u + I*v`, eliminate `v` to get an equation for `u` and then separate the real roots of that polynomial. Some of those may be multiple indicating that the original equation has several roots having the same real part.

skirpichev commented:
@jksuom, that looks promising.

It would be better if we will have a criteria that the original equation has several roots with same real part on a given interval.  But I think, the global criteria should work too.  It's all we need: if the count of "indistinguishable" (by real part) roots is exactly equal to the precomputed number of roots with same real part - we could stop the refinement process and sort the rest of "indistinguishable" roots by imaginary part.

skirpichev commented:
Hmm, I suspect that suggested by @jksuom method depends on ability to find real roots for polynomials with algebraic real coefficients.

smichr commented:
> Does your _imag_count() work on x**4 + 10*x**2 + 1?

yes
```
>>> _imag_count(Poly(x**4+10*x**2+1))
4
```

smichr commented:
> could you describe your algorithm?

The identification of roots for univariate, irreducible polynomials is the same as in master (pre my previous modifications) except that now 

1) complex roots are not sorted (but, by merit of construction (`roots.extend([root.conjugate(), root])`), always appear in pairs). There is no reason to sort. And without the ability to identify when non-real roots share the same real part, there is no way to do so. That the roots always appear in pairs has allowed the addition of the `_eval_conjugate` method to be added.

2) non-real roots are further refined if there are imaginary roots present so that the pure-imaginary can be identified unambiguously. This allows the `_eval_is_imaginary` routine to be able to give an answer where, in master, it was not previously possible without evaluation.

Regarding 1, you ask how they are identified:

> until either ax > 0 or bx < 0? Why do you believe that this process eventually ends? Could you prove it?

All roots with non-zero real part must eventually refine to have a positive or negative real part. If we know how many are not purely imaginary, we stop refining once they have been identified. They are identified by having both `ax*bx > 0`. Then the remaining root boundaries are those of the pure-imaginary bounds having `ax*bx <= 0`.

Regarding 2 you ask
> how it helps solving #14293

This is not a concern for me. Having the roots in conjugate pairs is sufficient.

> you (still) assume that pure imaginary roots are present only in special factors with two terms

No, you can see in the algorithm that I consider that imaginary roots of irreducible polynomials only come from the roots of `q(x)**2 +b*q(x) + c` . If `q(x)` == x the imaginary roots only occur when `b == 0` and `c > 0`. If `q(x)==x**n` (where `n` is an integer greater than 1) then there are two roots for each positive or negative root of `q(x)`. For `x**4 + 10*x**2 + 1` we have `q(x) == x**2` and we have two negative roots for `q(x)`. Each of these produces two imaginary roots for `x`. For x**6 + 10*x**3 + 1` where `q(x) == x**3` there would be no imaginary roots. Changing the 1 to -1 would give 2 imaginary roots; changing 1 and 10 to -1 and -10 would give 4 imaginary roots.

skirpichev commented:
Oh, I see, it seems you hardcoded some heuristic for polynomial with few terms, which have imaginary roots.  Unfortunately, you assume that there could no more that 3 terms.  Why do you think so, can you prove this intriguing assertion?  This is not true.

skirpichev commented:
Now for your second reply.

> complex roots are not sorted (but, by merit of construction (roots.extend([root.conjugate(), root])), always appear in pairs). There is no reason to sort. 

There are, that's was explained in #14293.  Yes, from rootisolation module they are coming in pairs and somehow sorted.  But it's not an ordering in the mathematical sense, e.g. first by real part and then break ties by imaginary part.  Next time some GSoC student might change arguments of dup_isolate_complex_roots_sqf or select different default isolation rectangle, or (improbably) implement some optimizations from Collins paper.  And all this will affect artificial sorting.

> This allows the _eval_is_imaginary routine to be able to give an answer where, in master, it was not previously possible without evaluation.

Ok, I see.  This is next "new feature" PR, not something related to the real issue.

> If we know how many are not purely imaginary, we stop refining once they have been identified. 

Agreed.  Unfortunately, you counting of imaginary roots is wrong.

> how it helps solving #14293 This is not a concern for me. Having the roots in conjugate pairs is sufficient.

Well, feel free to close issue.  Nice to see that arbitrary identification of roots is "not a concern" for sympy devs ;)

> the algorithm that I consider that imaginary roots of irreducible polynomials only come from the roots of q(x)**2 +b*q(x) + c ...

Ok, thank you for an explanation.  As I said before, you started from a wrong assumption.

skirpichev commented:
BTW, to find purely imaginary roots of p(x) you could just use dup_isolate_real_roots() for p(I*x).  That shouldn't be too expensive c.f. other alternatives.

Complex root isolation then must be adapted to use this information, namely: run isolation algorithms in left and right half-planes separately (AFAIR, only in the second case purely imaginary roots will be included).

As we already have real_roots(), so why not have imagiinary_roots()?  If this will not reduce performance, your pr definitely does make sense as a new feature.

skirpichev commented on sympy/polys/rootoftools.py:
```diff
@@ -271,135 +321,69 @@ def _reals_sorted(cls, reals):
-                break
+        # get the intervals pairwise-disjoint.
+        # If rectangles were drawn around the coordinates of the bounding
+        # rectangles, no rectangles would intersect after this procedure.
```
Oh, I bet they will...

smichr replied:
This is the task of `refine_disjoint`: refine the boundaries until either the x range or y range of self is distinct from other. When this condition is met the rectangles can no longer intersect. This does not mean that they will not overlap *when slid*, however. That's why being disjoint is not sufficient for sorting. (I am only guessing that that is what you are thinking is being claimed by this comment. It would help if you actually give an example so we have something concrete to talk about.)

skirpichev replied:
Oh, indead, I wrongly interpret this as being pairwise-disjoint in real part.  Sorry.

smichr commented:
> to find purely imaginary roots of p(x) you could just use dup_isolate_real_roots() for p(I*x)

I am aware of that but am not savvy enough in polynomial theory to know how to use it well. I am getting my education on the fly here.

This is a sketch of how I would use the method of testing for real roots of p(I*x):

```
if p(x) has all even powers then imaginary roots are the real roots of p(I*x)

if any power is odd, then p(I*x) == e(x) + I*o(x) and o(x) = x**m*q(x); imaginary roots of p(x) are
common roots of e(x) and q(x); we disregard the trivial root of x = 0 for x**m since that is not
imaginary. 

    If the degree of p(x) is even, e(x) cannot equal q(x) since the degree of q(x) is necessarily
    less than e(x)

    If e(x) == q(x) the imaginary roots of p(x) are the real roots of e(x) otherwise

    the real roots of the factors in common between e(x) and q(x) are the imaginary roots of p(x)
```

skirpichev commented:
I assume, you are talking about p(x) over ZZ/QQ?  If, so, yes.  This should work.

In case of the Diofant you can just convert domain to gaussian rationals, take composition of p(x) and q(x)=I*x, i.e. p(I*x) - then just get real roots of this polynomial, if any.  See diofant/diofant#625.

smichr commented:
> Nice to see that arbitrary identification of roots is "not a concern" for sympy devs ;)

I think it's an interesting concern but I have a very limited mathematical background for addressing this issue. (Is there even a consensus on how to refer to roots of a polynomial by index?) I am trying to make it so the RootOf instances are well ordered (so conjugate will work) and that they will evaluate without hanging.

skirpichev commented:
> I am trying to make it so the RootOf instances are well ordered (so conjugate will work)

``conjugate()`` may be implemented to work for any ordering.

smichr commented:
> conjugate() may be implemented to work for any ordering

yes. I set up a method wherein the conjugate was searched for within the complexes but I like the relationship that a pair is `RootOf(f, i), RootOf(f, i + 1)` for `i in range(nreal, degree)`

smichr committed:
improve identification of imaginary

smichr committed:
ay already has right sign

smichr committed:
no corner check

smichr committed:
watch out for non-symbol generators

smichr commented:
@skirpichev , I formalized a question [here](https://math.stackexchange.com/questions/2687963/are-there-purely-imaginary-roots-of-an-irreducible-whose-exponents-are-mixed-par) regarding counting of imaginary roots for certain types of polynomials.

smichr committed:
simplify _imag_count

See https://math.stackexchange.com/questions/2687963/are-there-purely-imaginary-roots-of-an-irreducible-whose-exponents-are-mixed-par

smichr committed:
use cached roots when possible

smichr commented:
Are you able to give this a review, @jksuom ?

smichr committed:
make custom dict to get/set root cache

It's nice to see the docstring values back to their
old values as in b1796095f8...but I am not sure what was
slipping by into the reals/complexes cache (perhaps) as a non-PurePoly
to cause that to happen. (It wasn't the change made that checked that
the imaginary root had an imaginary part: I tried making that change
alone and the doctests did not change to the values committed here.)

smichr commented:
If tests pass and there is no more feedback I will commit this in about 2 days.

jksuom commented:
I seem to have missed this, I'm sorry. I'll try to look into the code tomorrow.

smichr commented:
I have to check the evalf routine. It's losing track of the root somehow:
```
>>> rootof(x**5 - 5*x + 12, 1).n()
-0.351854240827372 - 1.70956104337033*I
>>> rootof(x**5 - 5*x + 12, 1).n(20)
-1.8420859661902543827 - 4.6365076883592767322e-69*I
>>> rootof(x**5 - 5*x + 12, 1).n(40)
-0.3518542408273719995586771904608686060769 - 1.709561043370328882014081953239973285284*I
```

smichr committed:
get non-real intervals off of real axis

smichr commented:
I am stymied by this behavior which happens when I try to get better control of the cached intervals:

```
This is ok
>>> id(rootof(x**5 - 5*x + 12, 1))
51487296
>>> id(rootof(x**5 - 5*x + 12, 1))
51487296

This is not (and there is an existing test for this error)
>>> id(rootof(x**5 - 5*x + 12, 1)._get_interval())
50256368
>>> id(rootof(x**5 - 5*x + 12, 1)._get_interval())
51267184
```

smichr commented:
In e14a5918 Ondrej noted:

```
    Do not store the refined interval in RootOf
    
    This fixes a caching bug. A test has been added, that failed before.
    What happened is that the evalf() method refines the interval and stores it
    back into the RootOf() method, which uses global cache... Subsequent calls to
    RootOf with the same polynomial and the *same* root index then assigns a wrong
    interval for the root.
    
    Ideally, the RootOf() implementation should be improved so that it is robust
    against refining the intervals. For the time being, this patch fixes the issue.
```

This is how it behaved just before that commit:

```
>>> RootOf(x**5 - 5*x + 12, 1).n()
-0.351854240827372 - 1.70956104337033*I
>>> RootOf(x**5 - 5*x + 12, 1).n()
-0.351854240827372 + 1.70956104337033*I  <--- positive imaginary part
>>> RootOf(x**5 - 5*x + 12, 1).n()
-0.351854240827372 - 1.70956104337033*I
>>> RootOf(x**5 - 5*x + 12, 1).n()
-0.351854240827372 - 1.70956104337033*I
>>> RootOf(x**5 - 5*x + 12, 1).n()
-0.351854240827372 - 1.70956104337033*I
>>> RootOf(x**5 - 5*x + 12, 1).n()
-0.351854240827372 - 1.70956104337033*I
>>> RootOf(x**5 - 5*x + 12, 1).n()
-0.351854240827372 - 1.70956104337033*I
>>> RootOf(x**5 - 5*x + 12, 1).n()
-0.351854240827372 - 1.70956104337033*I
>>> RootOf(x**5 - 5*x + 12, 1).n()
-0.351854240827372 - 1.70956104337033*I
>>> RootOf(x**5 - 5*x + 12, 1).n()
-0.351854240827372 - 1.70956104337033*I
>>> RootOf(x**5 - 5*x + 12, 1).n()
-0.351854240827372 - 1.70956104337033*I                                                                          
```

Now it doesn't have that problem so I wonder if the test can be dropped (or another method added so `==` works, `_eval_Eq`?).

```
>>> set([RootOf(x**5 - 5*x + 12, 1).n() for i in range(10)])
{-0.351854240827372 - 1.70956104337033*I}
```

jksuom commented:
I find the sorting of complex roots hard to understand. Perhaps this is not the final form?

smichr commented:
> I find the sorting of complex roots hard to understand.

There is no sorting happening there (yet [1]), only an affirmation of orderliness: the conjugates must follow one another and the roots of a given factor must be contiguous

[1] sorting can't happen until we can be sure that roots that have the same real part can be identified. And my preference (as I said in #14293) is to not sort unless there is some need for it. And at present, I am not aware of any.

smichr committed:
args and __eq__ to ComplexInterval

smichr committed:
simplify _pure_factor_dict

smichr committed:
watch for troublesome convergence

I'm not sure if the refinement is working properly.
The added test has terrible convergence to the root.
I actually never saw it succeed.

smichr committed:
make nroots sorting the same as RootOf

smichr committed:
watch for ints ending with L for python 2.7

smichr commented:
The refinement is fine on the test case. There are two issues, however: 1) the evalf code does not have an exit for when the boundary width falls below an acceptable precision and 2) the function must have an incredibly small convergence radius for the method being used. It keeps going out of the bounds with the intitial guess that is given. But solving for the root as a real value works...this time. I'm not sure the method will not have similar problems for some other equation.

Also, the mpmath polyroots (I think that is the name) routine is so fast compared to evalf that I wonder whether it's just better to make the evalf method just calculate all the roots and use the intervals to make sure the right one is being returned. Otherwise, an implementation of [this algorithm](http://www.math.cornell.edu/~hubbard/NewtonInventiones.pdf) (if it's not already being used in mpmath) seems like a great option from my initial read. But maybe that's also better for mpmath since it's numerics related.

Unrelated note: I would expect (after processing the poly the first time) that subsequent commands like this would be fast, but they aren't. It takes about 13s on my machine to return after this line:

```
RootOf(x**16 + 32*x**14 + 508*x**12 + 5440*x**10 +
        39510*x**8 + 204320*x**6 + 755548*x**4 + 1434496*x**2 +
        877969, 10)._get_interval()
```
- [ ] figure out why

smichr committed:
don't use .get; update interval after _evalf

smichr committed:
give x0 and x1 for initial guess

The default x1 value is x0 + .25. That's probably why the
convergence was so bad: if the interval was smaller than .25,
the x1 value was set outside of the interval. Now the 0.25 is
scaled to the size of the interval and an explicit value of
x1 is given.

Here is another method that works but unless the degree of the
polynomial and the precision is low -- the product of the two--
then it can take a long time to compute all of the roots. So
even though assignment is fast, computing all roots is slow.

def _eval_evalf(self, prec):
    roots = self.poly.nroots(prec)
    i = self._get_interval()
    if self.is_real:
        a, b = map(lambda i: Rational(str(i)), (i.a, i.b))
        for r in roots:
            if a <= r <= b:
                return r
    ax, bx = i.ax, i.bx
    ay, by = i.ay, i.by
    ax, bx, ay, by = map(lambda i: Rational(str(i)), (ax, bx, ay, by))
    for r in roots:
        X, Y = r.as_real_imag()
        if ax <= X <= bx and ay <= Y <= by:
            return r
    raise NotImplementedError('method did not find corresponding root')

smichr committed:
use tol for findroot

Without tol, the routine goes further than necessary.

smichr committed:
use muller for complex roots

smichr committed:
use refine with tolerance

smichr commented:
note to self:
```
@mattpap said:
Simplified and fixed complex indexing scheme in RootOf
RootOf doesn't handle conjugate pairs in any specific way anymore, which greatly simplifies indexing scheme of complex roots and makes it possible to add support for Gaussian rational coefficients in future, when complex root isolation over this domain will be supported. This change makes evalfO of RootOf objects slower in the complex case, because now z and conjugate(z) are evaluated separately (previously conjugate(z) was trivially computed from z).
```

smichr committed:
hybrid approach

smichr committed:
set dx to 0 for imaginary; no get for _reals_cache

smichr committed:
make sure polys kw is handled

smichr committed:
update evaluation methods

smichr commented on sympy/integrals/tests/test_quadrature.py:
```diff
@@ -466,7 +466,7 @@ def test_jacobi():
             '0.95949297361449739']
     assert [str(r) for r in w] == [
-            '0.090675770007435371',
+            '0.090675770007435372',
```
confirmed with:
```
>>> gauss_jacobi(5, -S.Half, S.Half, 17)[1][0]
0.090675770007435372
>>> gauss_jacobi(5, -S.Half, S.Half, 25)[1][0]
0.09067577000743537155596027
```

smichr commented on sympy/polys/specialpolys.py:
```diff
@@ -62,7 +62,7 @@ def swinnerton_dyer_poly(n, x=None, polys=False):
         for i in range(2, n + 1):
             p = nextprime(p)
             a.append(sqrt(p))
-        return minimal_polynomial(Add(*a), x, polys=args.get('polys', False))
```
there was not 'args' argument

smichr commented on sympy/polys/rootoftools.py:
```diff
@@ -658,49 +698,139 @@ def _eval_evalf(self, prec):
+        # all kwargs are ignored
+        return self.eval_rational(n=prec_to_dps(prec))._evalf(prec)
+
+    def eval_rational(self, dx=None, dy=None, n=15):
```
This now works for non-real roots and takes absolute (dx, dy) error bounds or a precision/relative bound (n).

smichr committed:
add _reset method to CRootOf

smichr committed:
misc edits

smichr committed:
use _mpq_ until .p and .q can be used

smichr commented on sympy/polys/rootoftools.py:
```diff
@@ -711,94 +716,118 @@ def _n(self, prec):
-            if self.is_imaginary:
-                expr = self.expr.subs(g, I*g)
-            func = lambdify(g, expr)
+            rtol = S(10)**-(n + 1)  # +1 for guard digit
```
I'm not so sure the thinking about the guard digit is sound. Consider the number 0.339981043584856xx48. Whether xx -- the 16th and 17th decimal place -- will remain xx or be rounded up to x(x+1) (e.g. 13 -> 14) depends on what xx is and what it is that follows xx. When 47 follows xx, none fail. When 48 follows xx the following values for xx fail:
```
>>> for j in range(100):
...  i = S('0.339981043584856%02d48'%j).n(17)
...  if str(i)[-2:]!=str('%02d'%j):print(j)
...
13
30
47
99
```
Removing the first decimal digit so xx fall at positions 15 and 16 give a different set of xx numbers that fail when trying to evaluate at the last xx digit (in this case, 16):
```
>>> for j in range(100):
...  i = S('0.39981043584856%02d48'%j).n(16)
...  if str(i)[-2:]!=str('%02d'%j):print(j)
...
1
6
11
16
18
23
28
33
40
45
50
57
62
67
74
79
84
91
96
```
It seems like the only way to make sure the last digit is correct is to keep evaluating 1 more decimal place until there is no change.

smichr replied:
5 binary guard digits at most are needed to code each decimal digit, so we need at most n+2 extra decimal digits to ensure that the nth digit is correct. e.g. prec_to_dps(dps_to_prec(10)+5) -> 12 so 12 digits will be needed to capture the 10th digit correctly.

smichr committed:
use 2 guard digits

smichr committed:
Rational sympifies non-Basic p

An mpq converter was added to sympy converters and Rational
now attempts sympification of non-string arguments. This allows
PythonRational and mpmath.mpq, for example, to be converted to
Rational without special testing in other places of the codebase.

smichr commented:
ping @jksuom

smichr committed:
allow is_disjoint to compare re and cmplx intervals

smichr commented:
ping @jksuom -- do you have reservations about committing this as is?

jksuom commented:
I forgot this for a while, sorry. I'll try to review in a couple of days.

jksuom commented on sympy/polys/rootisolation.py:
```diff
@@ -1835,14 +1982,24 @@ def conjugate(self):
+    def _set_imaginary(self):
+        a1, a2 = self.a
+        b1, b2 = self.b
+        a = a1 - a1, a2
```
Will this be `0, a2`?

smichr replied:
It will be a `PythonRational(0)`, I believe. I was just avoiding creating it explicitly since PythonRational already knows how to do subtraction.

jksuom replied:
I thought so. It could probably also be GMPYRational. How about adding a comment?

jksuom commented on sympy/polys/tests/test_rootoftools.py:
```diff
@@ -244,7 +250,23 @@ def test_CRootOf_evalf():
+        n = ri.n(2)
+        assert str(i) != str(ri._get_interval())
+        ri._reset()
+        # XXX using strings because intervals don't compare equal
```
Could `__eq__` be redefined to satisfy this?

smichr replied:
I can try. I'm not sure why the equality fails as it is.

smichr committed:
define RealInterval args; Intervals have func

smichr committed:
explain computation of zero in _set_imaginary

jksuom commented on sympy/polys/rootoftools.py:
```diff
@@ -29,22 +29,84 @@
+        if not isinstance(k, PurePoly):
+            assert len(k.free_symbols) == 1
+            k = PurePoly(k, expand=False)
+        return self.__dict__[k]
```
It seems that `__dict__` is not the place where a dict object stores its data.
```
>>> from sympy.polys.rootoftools import _pure_key_dict
>>> from sympy.abc import x
>>> P = _pure_key_dict()
>>> P[x] = 1
>>> x in P
False
>>> P.items()
[]
```
If we want to use other dict methods than `__setitem__` and `__getitem__`, we should probably either define them explicitly or else try to use here something like `dict.__getitem__(self, k)`.

jksuom replied:
It is probably not necessary to make lightweight cache a subclass of dict, if its methods are not needed. It seems to me that there could be `self.cache = {}` in `__init__(self)` and `__setitem__(self, k, v)` could then use `self.cache[k] = v`.

smichr replied:
LOL. I meant "caveat emptor" not "laissez-faire"!

smichr committed:
_is_imaginary correction; caveat for _pure_key_dict

smichr commented:
- [ ] fix this: perhaps trying to update the intervals is causing a mis-calculation of the roots

```
>>> roots4 = Poly(x**4 + 10*x**2 + 1).all_roots()
>>> [i.n(2) for i in roots4]
[-2.5*I, 2.5*I, -5.0*I, 5.0*I]
>>> nroots(x**4 + 10*x**2 + 1, 2)
[-0.32*I, 0.32*I, -3.1*I, 3.1*I]
>>> roots4[0].poly(.32*I)  # plausible root
-0.0135142400000001
>>> roots4[0].poly(2.5*I)  # definitely wrong
-22.4375000000000
```

smichr committed:
remove _set_imaginary -- leave interval args alone

The a and b part of the Interval args are not the
only args affected by refinement. So unless you
know what you are doing, you should not modify
individual args of the interval objects.

smichr committed:
store root info in dict attribute

To minimize unexpected results, _pure_key_dict derives from
object so trying to do more than is defined will fail.

jksuom commented on sympy/polys/rootoftools.py:
```diff
@@ -29,22 +29,98 @@
-_reals_cache = {}
-_complexes_cache = {}
+
+class _pure_key_dict(dict):
```
`object` would suffice.

smichr replied:
I added the `__contains__` method.

skirpichev replied:
Can't you just use a different key, common both for Poly and PurePoly?

smichr committed:
remove setdefault and test

jksuom commented:
I think this looks good.

smichr committed:
test eval methods

smichr committed:
correct RootSum test

smichr committed:
14615: take integral extraction in powsimp

smichr commented:
I will commit along with a fix to #14615 since I am the one that wrote that code. It fixes the identified problem and doesn't break anything.

smichr commented:
Thanks for sticking through with this, @jksuom . I appreciate your careful eye and useful direction.
